### DISCUSSION TITLE
Vaibhav's Review

### DISCUSSION THREAD

**Discussion by vaibhav-alphaics** [boris-kz/CogAlg#75] [General]
Hi Boris and Chris,

Thank you for giving me a chance to be a part of this. I have always wanted to work on something for unsupervised learning as our mainstream approaches are highly dependent on labels to learn. I find your approach very interesting and complex at the same time. So I have been through your wiki and the code of line_patterns.py. I had multiple questions regarding the approach and the implementation. Some of them were already answered in the other discussions. So I will continue with the remaining ones:
1. I didn't understand the concept of feedback filters in your case. What information will they backpropagate and what parameters would be updated?
2. Do you think this algorithm could be generalized for different sets of problems? As many hyperparameters are involved, would they be consistent? Also, I would like to know how did u decide these values( For example: ave)
3. And what part exactly is under development that I can contribute to? Just so I can look at the code from that point of view. (Any suggestions to go through the code are also welcome.)

Problems faced in analyzing the code:
The code is a bit too much to understand for any new person to understand. I think we can improve it by adding more comments about each step and while defining all the variables. Also, we could create write the algorithm separately for anyone to understand without actually going through the code.

Thanks. I am excited to learn more and contribute to this algorithm.

**Comment by boris-kz** [comment]
Thanks Vaibhab!

1: Please see my reply here: https://github.com/boris-kz/CogAlg/discussions/59#discussioncomment-850227

2: Generalization is a reduction. In 1D alg (starting from line_patterns), we reduced input complexity to a single variable ordered in a single dimension. I don't know how to make it any more general than that. In this case the variable is brightness and the dimension is width, but "pixels" could be quantized input of any modality in any dimension.   
It does have to be a raw sensory modality though, anything else is some form of second-hand encoding.

Hyperparameters are initialized average evaluated derivatives: educated guesses, to be updated by feedback, see 1). Results all comparisons are evaluated for clustering into patterns, as deviations from corresponding ave.

3: We are currently returning to 1D alg, after a spell in 2D alg. It's better to work-out core operations. So, current focus is on line_Pparams_draft, please see recent commits for progress. But you need to understand line_patterns first.

Sorry about complexity, I don't know how to make it any simpler. Yes, we can add more comments, you can ask inline questions where unclear. But all parameters, except for pixel__, are derived by operations in code. Those operations are the definitions. You do need memorize those, although more comments should help.

I don't think there is a better way to describe this alg than in Python, beyond verbal "outline of my approach" in Readme, and then docstrings in each module.
It is inherently complex, not because I like to complicate things, quite the opposite.
 
Thanks for thoughtful questions!
